transport/ProbeServer.cpp

```diff
@@ -0,0 +1,249 @@
+    if (it != _udpPorts.end())
+    {
+        return;
+    }
+
+    if (!_iceConfig.publicIpv4.empty() && endpoint.getLocalPort().getFamily() == AF_INET)
+    {
+        auto address = _iceConfig.publicIpv4;
+        address.setPort(port);
+        _probeCandidates.emplace_back(ice::TransportType::UDP, address);
+
+        _udpPorts.insert(port);
+
+        logger::debug("Registered UDP endpoint %s:%u", _name, address.ipToString().c_str(), port);
+    }
+
+    if (!_iceConfig.publicIpv6.empty() && endpoint.getLocalPort().getFamily() == AF_INET6)
+    {
+        auto address = _iceConfig.publicIpv6;
+        address.setPort(port);
+        _probeCandidates.emplace_back(ice::TransportType::UDP, address);
+
+        _udpPorts.insert(port);
+
+        logger::debug("Registered UDP endpoint %s:%u", _name, address.ipToString().c_str(), port);
+    }
+}
+
+void ProbeServer::onUnregistered(Endpoint& endpoint)
+{
+    if (endpoint.getTransportType() == ice::TransportType::UDP)
+    {
+        auto port = endpoint.getLocalPort().getPort();
+        _udpPorts.erase(port);
+
+        std::remove_if(_probeCandidates.begin(), _probeCandidates.end(), [port](const transport::ProbeCandidate& c) {
+            return c.transport == ice::TransportType::UDP && c.address.getPort() == port;
+        });
+    }
+}
+
+// ServerEndpoint::IEvents
+void ProbeServer::onServerPortRegistered(ServerEndpoint& endpoint)
+{
+    auto port = endpoint.getLocalPort().getPort();
+    auto it = _tcpPorts.find(port);
+    if (it != _tcpPorts.end())
+    {
+        return;
+    }
+
+    if (!_iceConfig.publicIpv4.empty() && endpoint.getLocalPort().getFamily() == AF_INET)
+    {
+        auto address = _iceConfig.publicIpv4;
+        address.setPort(port);
+        _probeCandidates.emplace_back(ice::TransportType::TCP, address);
+
+        _tcpPorts.insert(port);
+
+        logger::debug("Registered TCP server endpoint %s:%u", _name, address.ipToString().c_str(), port);
+    }
+
+    if (!_iceConfig.publicIpv6.empty() && endpoint.getLocalPort().getFamily() == AF_INET6)
+    {
+        auto address = _iceConfig.publicIpv6;
+        address.setPort(port);
+        _probeCandidates.emplace_back(ice::TransportType::TCP, address);
+
+        _tcpPorts.insert(port);
+
+        logger::debug("Registered TCP server endpoint %s:%u", _name, address.ipToString().c_str(), port);
+    }
+}
+
+void ProbeServer::onServerPortUnregistered(ServerEndpoint& endpoint)
+{
+    // Re-register to keep listening on probes;
+    // TcpServerEndpoint unregisters the listener after dispatching onIceTcpConnect
+    endpoint.registerListener(_probeCredentials.first, this);
+}
+
+void ProbeServer::onIceTcpConnect(std::shared_ptr<Endpoint> endpoint,
+    const SocketAddress& source,
+    const SocketAddress& target,
+    memory::UniquePacket packet)
+{
+    if (endpoint->getTransportType() == ice::TransportType::TCP)
+    {
+        replyStunOk(*endpoint, source, std::move(packet));
```

Comment:
We talked about the re-register of ufrag in tcp server, right?
I think we concluded that the ufrag must be registered again with the tcp server here since TCPserver thinks this is the final connection for that ufrag and automatically unregistered the ufrag.

Reply:
Yes, I re-register in `onServerPortUnregistered `callback.

`TcpServerEndpoint `removes listener for the `ufrag` after `onIceTcpConnect ` is called. Adding listener here wouldn't work because of duplicated keys (ufrag) as I understand.

Reply:
ok